docs/compilers/Design/Parser.md

```diff
@@ -15,14 +15,26 @@ The syntax model defines the contract between syntax and semantic analysis. If t
 
 There may be reasons to violate this rule, for example where the language is specified to be sensitive to context. For example, `await` is treated as a keyword if the enclosing method has the `async` modifier. 
 
+#### Impact of not following these DOs/DON'Ts
+
+The reasons for the above DOs/DON'Ts are not about perceived purity of the syntax or parser designs.  Rather, they have significant impact on both maintainability and correctness of the compiler and on the perf higher up parts of the stack that directly consume the syntax model (with the IDE being the primary consumer here).  Specifically:
+
+1. The presence of ambient context greatly impacts the ability to do incremental parsing properly.  Ambient context must be tracked in some fashion and incremental reuse across disparate contexts leads to violating the invariant that incremental reparsing produces the same tree as normal parsing.  This can and lead to fundamental brokenness that is only solved by a host restart.  As an example, consider an edit that adds 'async' to a method.  This must necessarily affect reuse of nodes within the method as they may be parsed separately.  Every piece of ambient context can have this effect and it dramatically makes it harder to reason about incremental edits and can lead to subtle and hard to diagnose or repro incremental errors in the wild.
+
+2. The presense of diagnostics beyond just `"missing token" and "unexpected token"` has impact again to the incremental parser.  Incremental parsing cannot reuse nodes that have errors in them, as it does not know what the root cause of the error was and if it would be fixed by an edit in a disparate part of the file.  As such, any additional diagnostic forces complete reparsing of that construct and all parent nodes above it.  This adds excess time to incremental parsing and can cause higher memory churn as well as new nodes must be created.  Both of these problems directly impact downstream consumption in IDE scenarios.  In the IDE typeing latency is a high value SLA target we need to meet.  However, many features must operate within the typing window in a manner that is 'correct' wrt to the syntax tree.  This includes, but is not limited to featuers like brace-insertion as well as indentation.  Both of these must happen near-instantiously to the user.  And both need an up-to-date SyntaxTree to determine what to do properly.  As such, these trees are retrieved synchronously using incremental parsing, so as little extra work as possible must be done so we can fit all the remainder of the work into the time-slice available.  Additional diagnostics interferes with this, adding extra unnecessary CPU and causing memory churn.
```

Comment:
```suggestion
2. The presence of diagnostics beyond just `"missing token" and "unexpected token"` has impact again to the incremental parser.  Incremental parsing cannot reuse nodes that have errors in them, as it does not know what the root cause of the error was and if it would be fixed by an edit in a disparate part of the file.  As such, any additional diagnostic forces complete reparsing of that construct and all parent nodes above it.  This adds excess time to incremental parsing and can cause higher memory churn as well as new nodes must be created.  Both of these problems directly impact downstream consumption in IDE scenarios.  In the IDE typing latency is a high value SLA target we need to meet.  However, many features must operate within the typing window in a manner that is 'correct' wrt to the syntax tree.  This includes, but is not limited to features like brace-insertion as well as indentation.  Both of these must happen near-instantaneously to the user.  And both need an up-to-date SyntaxTree to determine what to do properly.  As such, these trees are retrieved synchronously using incremental parsing, so as little extra work as possible must be done so we can fit all the remainder of the work into the time-slice available.  Additional diagnostics interferes with this, adding extra unnecessary CPU and causing memory churn.
```